lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -27,6 +136,344 @@
 
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
-LIBPATCH = 1
+LIBPATCH = 2
+
+from ipaddress import IPv4Address, IPv4Network
+from typing import Dict, List, Union, TypeAlias
+from ops.framework import Object, EventSource, EventBase, ObjectEvents
+from ops.charm import CharmBase
+from ops import RelationChangedEvent, RelationDepartedEvent, Relation
+import logging, json
+
+logger = logging.getLogger(__name__)
+
+Network: TypeAlias = Dict[
+    str,  # 'network' | 'gateway' | 'routes'
+    Union[
+        IPv4Address,  # gateway, ex: '192.168.250.1'
+        IPv4Network,  # network, ex: '192.168.250.0/24'
+        List[  # List of routes
+            Dict[
+                str,  # 'destination' | 'gateway'
+                Union[
+                    IPv4Address,  # gateway, ex: '192.168.250.3'
+                    IPv4Network,  # destination, ex: '172.250.0.0/16'
+                ],
+            ]
+        ],
+    ],
+]
+
+RoutingTable: TypeAlias = Dict[
+    str,  # Name of the application
+    List[Network],  # All networks for this application
+]
+
+
+class RoutingTableUpdatedEvent(EventBase):
+    """
+    Charm event for when a host registers a route to an existing interface in the router
+    """
+
+    def __init__(self, handle, data=None):
+        super().__init__(handle)
+        self.data = data
+
+    def snapshot(self):
+        return {"data": self.data}
+
+    def restore(self, snapshot):
+        self.data = snapshot["data"]
+
+
+class RouterProviderCharmEvents(ObjectEvents):
+    routing_table_updated = EventSource(RoutingTableUpdatedEvent)
+
+
+class RouterRequirerCharmEvents(ObjectEvents):
+    routing_table_updated = EventSource(RoutingTableUpdatedEvent)
+
+
+def _validate_network(network_request: Network, existing_routing_table: RoutingTable) -> bool:
```

Comment:
The issue with returning a bool is the fact that there could be multiple different types of errors when validating the network, and that information is lost when returning. That's why I chose to name it `_validate_network` instead of `_network_is_valid` or other variations. The same issue is not present with _network_name_taken, because there could only be one reason why this function returns a False. 

I don't consider these two functions to have similar functionality because of this. If we do any change, I'd rather have `_network_name_taken` raise a ValueError than convert the other function to bool.